From fff9731e4dc0040315f4bdb8b9e5c86664d424fb Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 19 Apr 2020 11:43:17 +0200
Subject: [PATCH] coccinelle: adjust NULL check before free()
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The free() function checks if its argument is NULL. We should avoid
checking for NULL before calling free like in

    if (result->tds)
        free(result->tds);

The list of relevant functions differs between Linux and U-Boot, e.g. we
use free().

Adjust the list of relevant functions.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 scripts/coccinelle/free/ifnullfree.cocci | 24 +++++++++++-------------
 1 file changed, 11 insertions(+), 13 deletions(-)

diff --git a/scripts/coccinelle/free/ifnullfree.cocci b/scripts/coccinelle/free/ifnullfree.cocci
index 14a4cd98e8..2d59545d7d 100644
--- a/scripts/coccinelle/free/ifnullfree.cocci
+++ b/scripts/coccinelle/free/ifnullfree.cocci
@@ -1,10 +1,11 @@
+// SPDX-License-Identifier: GPL-2.0-only
 /// NULL check before some freeing functions is not needed.
 ///
 /// Based on checkpatch warning
 /// "kfree(NULL) is safe this check is probably not required"
 /// and kfreeaddr.cocci by Julia Lawall.
 ///
-// Copyright: (C) 2014 Fabian Frederick.  GPLv2.
+// Copyright: (C) 2014 Fabian Frederick.
 // Comments: -
 // Options: --no-includes --include-headers
 
@@ -18,21 +19,19 @@ expression E;
 @@
 - if (E != NULL)
 (
-  kfree(E);
+  free(E);
 |
-  kzfree(E);
+  kfree(E);
 |
-  debugfs_remove(E);
+  vfree(E);
 |
-  debugfs_remove_recursive(E);
+  vfree_recursive(E);
 |
-  usb_free_urb(E);
+  kmem_cache_free(E);
 |
   kmem_cache_destroy(E);
 |
-  mempool_destroy(E);
-|
-  dma_pool_destroy(E);
+  gzfree(E);
 )
 
 @r depends on context || report || org @
@@ -41,9 +40,8 @@ position p;
 @@
 
 * if (E != NULL)
-*	\(kfree@p\|kzfree@p\|debugfs_remove@p\|debugfs_remove_recursive@p\|
-*         usb_free_urb@p\|kmem_cache_destroy@p\|mempool_destroy@p\|
-*         dma_pool_destroy@p\)(E);
+*	\(free@p\|kfree@p\|vfree@p\|debugfs_remove_recursive@p\|
+*         kmem_cache_free@p\|kmem_cache_destroy@p\|gzfree@p\)(E);
 
 @script:python depends on org@
 p << r.p;
@@ -55,5 +53,5 @@ cocci.print_main("NULL check before that freeing function is not needed", p)
 p << r.p;
 @@
 
-msg = "WARNING: NULL check before freeing functions like kfree, debugfs_remove, debugfs_remove_recursive or usb_free_urb is not needed. Maybe consider reorganizing relevant code to avoid passing NULL values."
+msg = "WARNING: NULL check before some freeing functions is not needed."
 coccilib.report.print_report(p[0], msg)
-- 
2.25.1

